### PR TITLE
Multiple time integrators in an input

### LINKED ISSUE #19228: Crank-Nicolson totally wrong if any PDE in the set missing a time derivative term
## Bug Description
If a PDE does not have a time derivative (say the mass equation for incompressible Navier-Stokes), then you will end up solving, for that sub-vector of dofs, the equation: residual + residual_old = 0. In my personal experience this yields solutions that look like this in time:
Screen Shot 2021-10-27 at 1 01 36 PM

## Steps to Reproduce
Solve a system of equations in which at least one equation does not have a time derivative. Your solution is in for a literally bumpy ride

## Impact
Makes users distrust MOOSE

### PR DESCRIPTION
**Pull request by lindsayad**
Closes #19228

### CODE REVIEW

**Comment by lindsayad** [comment]
relies on https://github.com/libMesh/libmesh/pull/3856 and https://github.com/libMesh/libmesh/pull/3857

**Comment by lindsayad** [comment]
don't close

File: framework/include/systems/SystemBase.h [hunk 1]
```diff
@@ -950,6 +943,28 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
...
+   * Retrieve the time integrator that integrates the given variable's equation. If no suitable time
+   * integrator is found (this could happen for instance if we're solving a non-transient problem),
+   * then a nullptr will be returned
+   */
+  const TimeIntegrator * getPossiblyNullTimeIntegrator(const unsigned int var_num) const;
+
+  /**
+   * @returns All the time integrators owned by this system
+   */
+  const std::vector<std::shared_ptr<TimeIntegrator>> & getTimeIntegrators();
```
Comment (loganharbour) [inline framework/include/systems/SystemBase.h:966]: Does it make sense here to not grant write access? i.e., reconstruct and give a `std::vector<std::shared_ptr<const TimeIntegrator>>`

File: framework/include/timeintegrators/AStableDirk4.h [hunk 2]
```diff
@@ -62,6 +62,7 @@ class AStableDirk4 : public TimeIntegrator
                                 ADReal & ad_u_dotdot) const override;
   virtual void solve() override;
   virtual void postResidual(NumericVector<Number> & residual) override;
+  virtual bool overridesSolve() const override { return true; }
```
Comment (loganharbour) [inline framework/include/timeintegrators/AStableDirk4.h:65]: Another non-required change, but I think we do this more now via a private param in the base class

File: framework/include/timeintegrators/TimeIntegrator.h [hunk 3]
```diff
@@ -164,6 +179,12 @@ class TimeIntegrator : public MooseObject, public Restartable
    */
   unsigned int getNumLinearIterationsLastSolve() const;
 
+  /**
+   * Copy from one vector into another. If the time integrator has been restricted to a subset of
+   * variables, then this will selectively copy their dofs
+   */
+  void copyVector(NumericVector<Number> & from, NumericVector<Number> & to);
```
Comment (loganharbour) [inline framework/include/timeintegrators/TimeIntegrator.h:186]:
```suggestion
  void copyVector(const NumericVector<Number> & from, NumericVector<Number> & to);
```

File: framework/src/restart/DataIO.C [hunk 4]
```diff
@@ -326,7 +326,11 @@ dataStore(std::ostream & stream,
           std::unique_ptr<libMesh::NumericVector<Number>> & v,
           void * context)
 {
-  mooseAssert(v, "Null vector");
+  bool have_vector = v.get();
```
Comment (loganharbour) [inline framework/src/restart/DataIO.C:329]: Can you add a comment here as to why you needed this? As it's a bit inconsistent with other nullptr things

File: framework/src/timeintegrators/AStableDirk4.C [hunk 5]
```diff
@@ -92,7 +92,9 @@ AStableDirk4::computeTimeDerivatives()
   u_dot = *_solution;
   computeTimeDerivativeHelper(u_dot, _solution_old);
   u_dot.close();
-  _du_dot_du = 1. / _dt;
+  for (const auto i : index_range(_du_dot_du))
```
Comment (loganharbour) [inline framework/src/timeintegrators/AStableDirk4.C:95]: Should there be a helper for this? Looks like it's used quite a lot

File: framework/src/timeintegrators/ImplicitEuler.C [hunk 6]
```diff
@@ -32,11 +32,26 @@ ImplicitEuler::computeTimeDerivatives()
...
-  computeTimeDerivativeHelper(u_dot, _solution_old);
-  u_dot.close();
+  if (!_var_restriction)
+  {
+    u_dot = *_solution;
+    computeTimeDerivativeHelper(u_dot, _solution_old);
+  }
+  else
+  {
+    auto u_dot_sub = u_dot.get_subvector(_local_indices);
```
Comment (loganharbour) [inline framework/src/timeintegrators/ImplicitEuler.C:42]: Same for this with a helper

File: framework/src/timeintegrators/TimeIntegrator.C [hunk 7]
```diff
@@ -33,26 +35,81 @@ TimeIntegrator::TimeIntegrator(const InputParameters & parameters)
...
+
+  _solution_sub = NumericVector<Number>::build(_solution->comm());
+  _solution_old_sub = NumericVector<Number>::build(_solution_old.comm());
+}
+
 void
 TimeIntegrator::solve()
 {
-  _nl.system().solve();
+  mooseError("Calling this method is no longer supported");
```
Comment (loganharbour) [inline framework/src/timeintegrators/TimeIntegrator.C:107]:
```suggestion
  mooseError("Calling TimeIntegrator::solve() is no longer supported");
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
Looks great. Most of this is just suggestions

File: framework/include/systems/SystemBase.h [hunk 8]
```diff
@@ -244,9 +244,9 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
    */
   virtual void addDotVectors();
 
-  virtual Number & duDotDu() { return _du_dot_du; }
+  virtual std::vector<Number> & duDotDus() { return _du_dot_du; }
   virtual Number & duDotDotDu() { return _du_dotdot_du; }
-  virtual const Number & duDotDu() const { return _du_dot_du; }
+  virtual const Number & duDotDu(unsigned int var_num = 0) const;
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/systems/SystemBase.h:249]: It is the exact same scenario. I am happy to make the change a breaking one in order to not have these defaulted arguments

File: framework/include/timeintegrators/AStableDirk4.h (hunk 2 above)
Comment (lindsayad) [reply framework/include/timeintegrators/AStableDirk4.h:65]: A private data member or private method? I want the pure virtual to ensure that developers must override it. I initially missed a time integrator in the solid mechanics module which led to a bunch of diffs when I didn't have the pure virtual

File: framework/include/systems/SystemBase.h (hunk 1 above)
Comment (lindsayad) [reply framework/include/systems/SystemBase.h:965]: Both `Transient` and `FEProblemBase` call non-const methods on the time integrators returned by this API. The vector should be constant, but the pointers shouldn't be

File: framework/include/timeintegrators/AStableDirk4.h (hunk 2 above)
Comment (loganharbour) [reply framework/include/timeintegrators/AStableDirk4.h:65]: I didn't realize it was a pure virtual. I like that more. Sounds good to me.

File: framework/include/systems/SystemBase.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (loganharbour) [reply framework/include/systems/SystemBase.h:249]: Up to you.

[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/systems/SystemBase.h:249]: I'll keep it the way it is for now, but I've opened https://github.com/idaholab/moose/issues/28966 to keep track of this and take care of it quickly after propagating the MOOSE changes into Griffin

**Review by loganharbour — APPROVED** [review]

**Comment by lindsayad** [comment]
thanks for the review!

**Pull request merged**